File: sdks/ruby/examples/direct_chat.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

# DIRECT (in-process) — llmux through the C ABI, no child process and no port.
#
#   ruby sdks/ruby/examples/direct_chat.rb
#
# Environment:
#   LLMUX_LIBRARY      path to libllmux.dylib/.so (default: see Ffi.resolve_library)
#   LLMUX_CONFIG_JSON  an llmux config document, inline. Unset means built-in
#                      defaults plus the environment (OPENAI_API_KEY, ...).
#   LLMUX_MODEL        model to ask (default: openai/gpt-4o-mini)
#
# Before you put this in a Rails app, read sdks/ruby/README.md: Unicorn and
# clustered Puma fork their workers, and the Go runtime inside libllmux does not
# survive fork(). examples/sidecar_chat.rb is the shape for those.

$LOAD_PATH.unshift(File.expand_path("../lib", __dir__))
require "llmux/ffi"

config = ENV["LLMUX_CONFIG_JSON"]
config = nil if config.nil? || config.empty?
model = ENV.fetch("LLMUX_MODEL", "openai/gpt-4o-mini")

# Ffi.open closes the handle however the block exits — a raise, a return, or a
# normal finish. It is `ensure` written once instead of at every call site.
Llmux::Ffi.open(config: config) do |llmux|
  puts "library : #{llmux.library_path}"
  puts "abi     : #{llmux.version}"

  # 1. A call that touches no upstream: the routing table itself.
  ids = llmux.call("models")["data"].map { |m| m["id"] }
  puts "models  : #{ids.size} (#{ids.take(3).join(', ')}#{ids.size > 3 ? ', …' : ''})"

  # 2. A unary chat completion.
  answer = llmux.call("chat",
                      model: model,
                      messages: [{ role: "user", content: "Say hello in five words." }])
  puts "chat    : #{answer.dig('choices', 0, 'message', 'content').strip}"

  # 3. The same request, streamed. The block runs on THIS thread, synchronously,
  #    before stream returns.
  print "stream  : "
  $stdout.flush
  chunks = llmux.stream("chat",
                        model: model,
                        messages: [{ role: "user", content: "Say hello in five words." }]) do |chunk, _raw|
    delta = chunk.dig("choices", 0, "delta", "content")
    next if delta.nil? || delta.empty?

    print delta
    $stdout.flush
  end
  puts "\n          (#{chunks} chunks)"

  # 4. Stopping early. Returning :stop stops the stream and is NOT an error.
  seen = 0
  llmux.stream("chat",
               model: model,
               messages: [{ role: "user", content: "Count to fifty." }]) do
    seen += 1
    seen >= 3 ? :stop : nil
  end
  puts "early   : stopped after #{seen} chunks, no error"

  # 5. The error path. The message llmux wrote to *err is freed with llmux_free
  #    inside call_raw before the exception is raised — an error is not a leak.
  begin
    llmux.call("summon-a-demon", {})
    puts "error   : UNEXPECTED — an unknown method should fail"
  rescue Llmux::Error => e
    puts "error   : #{e.message}"
  end

  # 6. What the GVL choice buys. fiddle releases the GVL for the call, so other
  #    Ruby threads keep running while llmux is busy.
  ticks = 0
  stop = false
  worker = Thread.new { ticks += 1 until stop }
  llmux.call("chat", model: model, messages: [{ role: "user", content: "hi" }])
  stop = true
  worker.join
  puts "threads : a background Ruby thread ran #{ticks} times during one call"
end

puts "closed  : ok"

File: sdks/php/examples/direct_chat.php
<?php

declare(strict_types=1);

/**
 * DIRECT (in-process) — llmux through the C ABI, no child process and no port.
 *
 *   php sdks/php/examples/direct_chat.php
 *
 * Environment:
 *   LLMUX_LIBRARY      path to libllmux.dylib/.so (default: see Ffi::resolveLibrary)
 *   LLMUX_CONFIG_JSON  an llmux config document, inline. Unset means built-in
 *                      defaults plus the environment (OPENAI_API_KEY, ...).
 *   LLMUX_MODEL        model to ask (default: openai/gpt-4o-mini)
 *
 * Before you copy this into a web application, read sdks/php/README.md. Under
 * php-fpm this is measurably the wrong shape: use sidecar_chat.php.
 */

require __DIR__ . '/../src/LlmuxException.php';
require __DIR__ . '/../src/Ffi.php';

use Llmux\Ffi;
use Llmux\LlmuxException;

$config = ($c = getenv('LLMUX_CONFIG_JSON')) !== false && $c !== '' ? $c : null;
$model = ($m = getenv('LLMUX_MODEL')) !== false && $m !== '' ? $m : 'openai/gpt-4o-mini';

try {
    $llmux = new Ffi($config);
} catch (LlmuxException $e) {
    fwrite(STDERR, "could not load libllmux: {$e->getMessage()}\n");
    exit(1);
}

// try/finally is the whole point: every path below — a thrown exception, an
// early return, a failed call — still closes the handle exactly once.
try {
    echo "library : {$llmux->libraryPath()}\n";
    echo "abi     : {$llmux->version()}\n";

    // 1. A call that touches no upstream: the routing table itself.
    $models = $llmux->call('models');
    $ids = array_map(static fn (array $m): string => $m['id'], $models['data']);
    printf("models  : %d (%s%s)\n", count($ids), implode(', ', array_slice($ids, 0, 3)),
        count($ids) > 3 ? ', …' : '');

    // 2. A unary chat completion.
    $answer = $llmux->call('chat', [
        'model' => $model,
        'messages' => [['role' => 'user', 'content' => 'Say hello in five words.']],
    ]);
    echo "chat    : ", trim($answer['choices'][0]['message']['content']), "\n";

    // 3. The same request, streamed. The callback runs on THIS thread,
    //    synchronously, before stream() returns — measured in ffi/ctest/smoke.c,
    //    not assumed — so there is nothing to synchronise and no GIL analogue
    //    to worry about.
    echo 'stream  : ';
    $chunks = $llmux->stream('chat', [
        'model' => $model,
        'messages' => [['role' => 'user', 'content' => 'Say hello in five words.']],
    ], static function (array $chunk): void {
        $delta = $chunk['choices'][0]['delta']['content'] ?? '';
        if ($delta !== '') {
            echo $delta;
            flush();
        }
    });
    echo "\n          ({$chunks} chunks)\n";

    // 4. Stopping early. Returning false from the callback stops the stream and
    //    is NOT an error — llmux returns 0 and writes no message.
    $seen = 0;
    $llmux->stream('chat', [
        'model' => $model,
        'messages' => [['role' => 'user', 'content' => 'Count to fifty.']],
    ], static function () use (&$seen): bool {
        return ++$seen < 3;   // false on the 3rd chunk
    });
    echo "early   : stopped after {$seen} chunks, no error\n";

    // 5. The error path. The message llmux writes to *err is freed with
    //    llmux_free inside callRaw() before the exception is thrown — an error
    //    is not a leak.
    try {
        $llmux->call('summon-a-demon', []);
        echo "error   : UNEXPECTED — an unknown method should fail\n";
    } catch (LlmuxException $e) {
        echo "error   : {$e->getMessage()}\n";
    }
} finally {
    $llmux->close();
}

echo "closed  : ok\n";
